### PR TITLE
Feature/reset password

### PR DESCRIPTION
**Pull request by RaulTsc**
Hey Liviu,

This is still a bit raw, I still got a few TODOs, such as

* Fix tests for change/reset password
* Add check for changepassword form for checking fields are equal

But I wanted to submit it a bit early to get feedback on it. 😄

### CODE REVIEW

File: src/server/controllers/index.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export function setupApiRoutes(app: express.Application): void {
   app.get('/api/users/me', authMiddleware, userController.getMe);
   app.post('/api/users/register', validate(userController.registerValidation), userController.registerUser);
   app.put('/api/users/resetpassword', validate(userController.resetPasswordValidation), userController.resetPassword);
-  app.put('/api/users/changepassword', authMiddleware, validate(userController.changePasswordValidation), userController.changePassword);
+  app.put('/api/users/changepassword', validate(userController.changePasswordValidation), userController.changePassword);
```
Comment (liviuignat) [inline src/server/controllers/index.ts]: I was thinking that this will be used when the user is logged in. When the user is authenticated I don't think he needs an email. Like an "Account Settings" form with:
```
oldPassword
newPassword
repeatPassword
```

But maybe we can change it in the future and use the same endpoint, if the user passes a valid token to the API (meaning he is logged in), he will be able to change the password directly by passing the old and the new password. What do you think?

File: src/server/controllers/userController.ts [hunk 2]
```diff
@@ -59,13 +61,13 @@ export async function resetPassword(req: Request, res: Response, next: NextFunct
...
-    const newPassword = randomstring.generate();
-
-    user.password = md5(newPassword);
+    const resetPasswordId = uuid.v1();
+    user.resetPasswordId = resetPasswordId;
     await updateUser(user);
 
     // Don't wait to send the email
-    sendUserResetPasswordEmail(user, newPassword);
+    const changePasswordUrl = req.get('Referrer').replace('resetpassword', `changepassword/${user.id}/${user.resetPasswordId}`);
```
Comment (liviuignat) [inline src/server/controllers/userController.ts]: What happens if this would be an API call done from some back-end system? Wouldn't the `Referrer` be `null`? Maybe we should put the `baseUrl` in `config`?

File: src/server/controllers/userController.ts [hunk 3]
```diff
@@ -59,13 +61,13 @@ export async function resetPassword(req: Request, res: Response, next: NextFunct
...
+    sendUserResetPasswordEmail(user, changePasswordUrl, language);
```
Comment (liviuignat) [inline src/server/controllers/userController.ts]: Actually even better, maybe the `changePasswordUrl` should be built inside `sendUserResetPasswordEmail`  and the `baseUrl` read from `config`.

File: src/server/controllers/index.ts (hunk 1 above)
Comment (RaulTsc) [inline src/server/controllers/index.ts]: Uhm, yeah. Makes sense. So then I'm thinking of two options:

1. Have same endpoint and then we can't use the authMiddleware here, we'd have to check the auth header manually in userController. Not sure if that can handled easily by passport.

2. Have two endpoints. Leave changepassword as is and use it for 'Account Settings' and then have a '/api/users/resetpasswordsteptwo' endpoint changes the password based on userId and resetPasswordId.

I personally like the second option better.

File: src/server/data/models/user.ts [hunk 4]
```diff
@@ -30,6 +30,9 @@ export default function defineUser(sequelize: Sequelize.Sequelize): IUserModel {
       type: Sequelize.STRING(256),
       field: 'last_name',
     },
+    resetPasswordId: {
+      type: Sequelize.UUID,
```
Comment (liviuignat) [inline src/server/data/models/user.ts]: I'm not sure if sequelize transforms this to snake case, usually I add the `field: 'reset_password_id'` just to be sure and not have surprises.

File: src/server/controllers/userController.ts (hunk 3 above)
Comment (RaulTsc) [inline src/server/controllers/userController.ts]: Yeah, totally agree.

File: src/server/controllers/userController.ts (hunk 2 above)
Comment (RaulTsc) [inline src/server/controllers/userController.ts]: Nice catch :D. It would probably be null. I agree, we should have the baseUrl in config and build the url in `sendUserResetPasswordEmail`

File: src/server/data/models/user.ts (hunk 4 above)
Comment (RaulTsc) [inline src/server/data/models/user.ts]: You're right, it doesn't. I'll add it.

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+          errorMessage={changePasswordError}
+          onSubmit={onSubmit}
+        />;
+
+    return (
+      <Paper className={css.ChangePasswordPage}>
+        <Helmet title="EverReal - change password" />
+        <h3><FormattedMessage id="ChangePasswordPage.page.title" /></h3>
+
+        {component}
```
Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: So, the page is always accessible via url, but if the route parameters are not correct, the user will get an error message when he tries to submit the form right?

File: src/server/controllers/index.ts (hunk 1 above)
Comment (liviuignat) [inline src/server/controllers/index.ts]: Yep, second option is way better, it's more verbose. What if we call the endpoints:
- `/api/users/me/changepassword` - this would be the old `/api/users/changepassword` where ou have to be authenticated
AND
- `/api/users/{userId}/changepassword` - and here you have to pass also the correct `{password, resetPasswordId}`. 

I have a feeling it's more restful like this, what do you think?

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx (hunk 5 above)
Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: Yes. He'll either get invalid userId or invalid resetPasswordId. I added a TODO which says we need to check both have valid UUID structure, because otherwise sequelize throws an error. Also we might want to show the error message in a div which replaces the two password fields (currently they are displayed under one of the fields).

Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: Maybe Joi can help with the validation, or a custom validator?

File: src/server/controllers/index.ts (hunk 1 above)
Comment (RaulTsc) [inline src/server/controllers/index.ts]: Sounds great!

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx (hunk 5 above)
Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: I'm sure Joi can do it, I've used it before to validate UUIDs.

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordForm.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          <FormTextField
+            withRef
+            ref="password"
+            fullWidth
+            disabled={isLoading}
+            name="password"
+            type="password"
+            hintText={<FormattedMessage id="App.password" />}
+            floatingLabelText={<FormattedMessage id="App.password" />}
+            validate={[required, repeatPasswordValidation]}
```
Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordForm.tsx]: @liviuignat I saw we repeatPasswordValidation as helper from er-common-components, but I'm not exactly sure how to use it.

Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordForm.tsx]: https://github.com/liviuignat/heimdall/blob/develop/src/universal/auth/pages/RegisterPage/RegisterForm.tsx#L64
```ts
repeatPasswordValidation(passwordFieldName: string): (value: any, formData: AllFormFields) => string | null
```

The repeatPasswordValidation returns a callback and it will look for the value to compare in `formData[passwordFieldName]`.

Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordForm.tsx]: `formData` is a JSON object with all the form fields: 
```js
{
  fieldName1: value1,
  fieldName2: value2
}
```

**Comment by liviuignat** [comment]
@RaulTsc ping me if you want me to do another review 😉

**Comment by RaulTsc** [comment]
Sure thing.

**Comment by RaulTsc** [comment]
Hey Liviu,

Please see the changes in the last commit. I want your feedback on two things in particular: `getComponent()` in `ChangePasswordPage.tsx` and tests for reset password (should we return the reset token as response so we can test it?).

Any other feedback is welcome aswell.

File: src/server/__tests__/user.test.ts [hunk 7]
```diff
@@ -67,48 +67,55 @@ describe('WHEN testing user endpoints', () => {
...
               expect(me.email).toEqual(createUserPayload.email);
             });
           });
         });
 
-        describe('WHEN resetting password for the new user', () => {
-          it('SHOULD return staus code 200', () => resetPasswordRequest(createUserPayload.email).expect(200));
+        // TODO: Fix this. Maybe check that resetPasswordId is generated. In order to check that the resetPasswordId
+        // is generated we must return it as response. We can then also test /{userId}/changepassword
+        // describe('WHEN doing resetpassword for a user', () => {
```
Comment (liviuignat) [inline src/server/__tests__/user.test.ts]: Shouldn't we leave at least the check for `status code 200`? What do you think?

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      </Paper>
+    );
+  }
+
+  private getComponent(css) {
```
Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: `render` method should be last. Interesting that `eslint` didn't complain. Usually it complains about the order of methods.

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+          <p><FormattedMessage id="ChangePasswordPage.label.errorMessage" /></p>
+          <div className={css.Links_container}>
+            <FormattedLink href="/resetpassword" className={css.Links_resetpassword}><FormattedMessage id="LoginPage.label.reset.password" /></FormattedLink>
+          </div>
+        </div>;
+    }
+
+    return component;
+  }
+}
```
Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: LGTM. I usually write like:
```js
const notStarted = !isChangePasswordSuccess && !changePasswordError;
const finishedWithError = isChangePasswordSuccess;
const finishedWithSuccess = !isChangePasswordSuccess && generalErrors.includes(changePasswordError);

return (
  <div>
    {notStarted &&  <ChangePasswordForm
      isLoading={isChangingPassword}
      errorMessage={changePasswordError}
      onSubmit={onSubmit}
    />}

    {finishedWithError &&  <div className={css.SuccessMessage}>
      <p><FormattedMessage id="ChangePasswordPage.label.successMessage" /></p>
        <div className={css.Links_container}>
          <FormattedLink href="/login" className={css.Links_login}><FormattedMessage id="LoginPage.label.login" /></FormattedLink>
        </div>
      </div>
    </div>}

    {finishedWithSuccess &&  <div className={css.ErrorMessage}>
      <p><FormattedMessage id="ChangePasswordPage.label.errorMessage" /></p>
      <div className={css.Links_container}>
        <FormattedLink href="/resetpassword" className={css.Links_resetpassword}><FormattedMessage id="LoginPage.label.reset.password" /></FormattedLink>
      </div>
    </div>}
  </div>
);
```

And, most of the time, I set `finishedWithError, finishedWithSuccess` on the reducer. The reducer know when the state changes and I think `notStarted === !finishedWithError && !finishedWithSuccess;`

Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: But that's also OK, it's up to you if you want to change or not. When the test is fixed, it's good to merge I think, but we need to make a decision with the `TODO` on the test.

File: src/server/__tests__/user.test.ts (hunk 7 above)
Comment (RaulTsc) [inline src/server/__tests__/user.test.ts]: I agree. We should just have the check for status code 200. We should not have it return the reset token because it might be a vulnerability, if someone finds out the user id for that email aswell they can reset a user's password.

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx (hunk 8 above)
Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: It actually complained about having private methods declared before public methods.

file: 'file:///Users/raul/Work/everreal/heimdall/src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx'
severity: 'Warning'
message: 'Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after public static fields. (member-ordering)'
at: '80,3'
source: 'tslint'

Should we change the eslint rule?

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx (hunk 9 above)
Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: I think your way of writing it is better, I'll do that aswell from now on. Will update & merge the PR.

Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: How about doing something like this on the reducer?
```
// reducer sets the visible state
case actionTypes.AUTH_CHANGE_PASSWORD:
  return Object.assign({}, state, {
    isChangingPassword: true,
    passwordChangedWithError: false,
    passwordChangedWithSuccess: false,
    changePasswordError: '',
  });
case actionTypes.AUTH_CHANGE_PASSWORD_SUCCESS:
  return Object.assign({}, state, {
    isChangingPassword: false,
    passwordChangedWithError: false,
    passwordChangedWithSuccess: true,
    changePasswordError: '',
  });
case actionTypes.AUTH_CHANGE_PASSWORD_FAIL:
  return Object.assign({}, state, {
    isChangingPassword: false,
    passwordChangedWithError: true,
    passwordChangedWithSuccess: false,
    changePasswordError: errorFomatter(action.error),
  });

// actual visible state
const notStarted = !passwordChangedWithError && !passwordChangedWithSuccess;
const finishedWithError = passwordChangedWithError;
const finishedWithSuccess = passwordChangedWithSuccess;
```

Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: I think it is not needed. We have `notStarted` as `isChangingPassword`.

What I would do in exchange is have two types of errors, `general` and `fields`(maybe use another name). Then, we would check against changePasswordError.type and not against an array of strings to determine what type of error it is.

Thoughts?

Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: I think it's a good idea, but I also think that we could do it in another PR.

Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: Agreed.

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx (hunk 8 above)
Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: This will be fixed as removed getComponent and moved the logic to the reducer. But this would still be an issue, maybe open a ticket?

**Comment by RaulTsc** [comment]
Inca o chestie importanta inainte de merge. O sa fac eu acum commit cu varianta finala si apoi o sa te rog pe tine sa traduci in germana.

File: src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx (hunk 8 above)
Comment (liviuignat) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: I think `private` can be easily changed to `public`. Also `tslint` does not use `airbnb` rules and maybe it should.

Comment (RaulTsc) [inline src/universal/auth/pages/ChangePasswordPage/ChangePasswordPage.tsx]: Yep. It can be changed. But I'd like to have `private` methods in the future.

**Comment by liviuignat** [comment]
Coming back to `private` vs `public` in components - I don't think it matters in components. There are a few reasons:
- they should have as less logic as possible, and the logic that they will have it's not really necessary to be "private"
- react is moving to pure functional components, so anyway that will not be a case anymore :)

**Comment by RaulTsc** [comment]
Yeah, but we're using the linter for the entire codebase. 

We just need to decide if we like having `private`s at the bottom vs changing the lint rule (I prefer the second option).

**Comment by liviuignat** [comment]
Ah, this is a rules that comes from typed languages. In Java or .NET it's a standard to have privates at the bottom and to have the "important" methods at the top. I would leave it for now and not use `private` in React components, but if it gives us problems we remove it in the future, what do you think?

**Comment by RaulTsc** [comment]
Cool.
